Merging wdl/improved_validation into master

mewthew opened the pull request:
This PR contains the Victor and Matthew's implementation of validation functions and Kate, Shriya, and Thor's design specification for sprints 1 and 2.

The wdl/improved_validation branch consists of code that validates YAML objects by generating lists of objects and attributes, which are then used for type-checking and printing. 

In our validate module, we have created functions such as: `list_type_check` and `list_print`, which allows us to iterate through a list of objects or attributes and call our singular `game_type_check`, `room_type_check`, and `item_type_check` functions that verify the types for a given game, room(s), or item(s). In our parse module, we have created functions that generates the lists used for validation, such as `extract_objects`.

The implementation for the two modules can be accessed within the wdl repository in the src directory and an example of our code can be accessed within the examples directory.

NOTE:
The Makefile within our repository currently does not link our library with our examples files yet, which we will be fixing ASAP (we've been given the go-ahead by the supervisors ;))

Within the docs directory, wdl.md includes the World Description Language specification that we developed. It contains definitions for terminology used throughout the game engine, syntax for the WDL file that the game designer will input, and detailed explanations and examples for the three main components of the WDL file (Game, Room, and Item). These objects all contain attributes that must be written in accordance with this specification. This includes proper indentation and appropriate usage of the "-" character as exemplified by the WDL object examples. We have also linked the documentation for available actions that the action_management team created in this document. There are other features being developed which will validate the WDL file that the game designer creates and input it into the structures for the game. We have designed this specification in relation to the current needs of the teams, but we will update the design as needed in future sprints.

This pull request reflects the work of:
Victor Almaraz-Argueta
Matthew Yang
Kate Lu
Shriya Bansal
Thor Anderskow

BIG THANKS TO OUR PR REVIEWERS, this could not be possible without you guys

Kate32815 commented:
![validation_review_p1](https://user-images.githubusercontent.com/30165182/58003037-83be7380-7aa5-11e9-907b-8c4ecac19776.jpg)
![validation_review_p2](https://user-images.githubusercontent.com/30165182/58003038-83be7380-7aa5-11e9-957c-9fd3e3bf04a5.jpg)
Sorry I wrote this out on paper instead of typing it! I can type it up if that's better

lydiafilipe commented:
@Kate32815 I would recommend typing comments and also typing them all on the code itself- when you add your review and hover over a line you see a little blue plus sign - click on that and you can add a comment for that line

jessxia commented on docs/WDL_README.md:
```diff
@@ -0,0 +1,44 @@
+# WDL Specification
+May 2, 2019
+
+###### A WDL file is a YAML file that has at least the three file objects: Game, Room, Objects.
```
Game, Room, and Items

mewthew replied:
We recently updated the most recent spec file found in chiventure/docs/wdl.md

jessxia replied:
woot

lydiafilipe commented on src/wdl/src/parse.c:
```diff
@@ -0,0 +1,60 @@
+        valid = list_check_type(ls, room_type_check);
+    else if (strcmp(str, attributes[1]))
+       valid = list_check_type(ls, item_type_check);
+    else
```
I would do another else-if comparison for players and have the else condition set valid=false - you shouldn't assume the input will always fall into the proper attributes

victor-a-a replied:
Addressed in commit `91e6223`

lydiafilipe commented on src/wdl/src/parse.c:
```diff
@@ -0,0 +1,60 @@
+/* see parse.h */
+id_list_t *extract_ids(attr_list_t *ls)
+{
+    return NULL;
```
Maybe leave  comment that this is a TODO right now

victor-a-a replied:
Addressed in commit `d94064e57a8b62bb184c96b855d94e7473f8a586`

lydiafilipe commented on src/wdl/examples/typecheck_demo.c:
```diff
@@ -0,0 +1,91 @@
+
+    // Verify game
+    bool game_check = game_type_check(game_obj);
+    if (game_check)
```
All if conditions, loops etc should have braces, even if they are one statement. Please add these in all your files/functions

victor-a-a replied:
Addressed in commit `91e6223`

mewthew commented:
Hi @Kate32815 ,
Thanks for all the feedback. Here are some of the responses/updates I made per your feedback:
1. To all your questions about optional fields, we plan on taking them into account this sprint (we were waiting for the spec to be completely finalized and agreed upon with game_state)
2. Fixed the comment concerning list_print()
3. Updated the comment concerning conditions_get_list()
4. Updated the comment concerning check_condition_attr()
5. Corrected the implementation for condition type_checking and printing
6. Corrected the documentation for check_connections()

Let me know if I missed anything!

mewthew commented on src/wdl/examples/typecheck_demo.c:
```diff
@@ -0,0 +1,91 @@
+    // Checks if extension is a .yaml extension
+    const char *ext = strchr(fname, '.');
+    const char *check = ".yaml";
+    if (strcmp(ext, check) != 0)
```
Got it. We will implement the strn functions after this pr.

victor-a-a commented on src/wdl/src/parse.c:
```diff
@@ -0,0 +1,60 @@
+{
+    char *attributes[3] = {"ROOMS", "ITEMS", "PLAYERS"};
+
+    if (strcmp(str, attributes[0]) != 0 &&
```
Addressed in commit `d94064e57a8b62bb184c96b855d94e7473f8a586`

victor-a-a commented on src/wdl/src/parse.c:
```diff
@@ -0,0 +1,60 @@
+    if (ls == NULL)
+        return NULL;
+
+    if (strcmp(str, attributes[0]))
```
Addressed in commit `d94064e57a8b62bb184c96b855d94e7473f8a586`

victor-a-a replied:
Addressed in commit `91e6223`

victor-a-a commented on src/wdl/README.md:
```diff
@@ -0,0 +1 @@
+
```
Addressed in commit `82eb80a`

victor-a-a commented:
Hi @lydiafilipe ,
Thank you for your feedback.  Here is a summary of the changes made:
1. Added content to our README.  More precisely, I made a note of decencies, added a basic description of each modules' function, and added a description of the `typecheck_demo` example's purpose/function.
2. Removed the array of strings for any function that had them.
3. Fixed the logic bug in `extract_objects()`.
4. Added braces to any conditionals or loops that lacked them.

lydiafilipe commented on src/wdl/src/validate.c:
```diff
@@ -0,0 +1,405 @@
+bool check_connection_attr(obj_t *obj)
+{
+    // verify types of fields
+    bool id = true, direction = true, through = true, conditions = true;
```
I notice throughout this file you have a lot of statements in the form of
if(boolean_expression) {
  boolean variable = false
}

This can actually be simplified by setting the boolean variable in terms of the boolean expression you want it to be based on. E.g. 
bool id = obj_get_type(obj, "to") == TYPE_STR;

Instead of 
bool id = true
if(obj_get_type(obj, "to") != TYPE_STR){
    id = false;
}

I think it would be cleaner to simplify all these statements

mewthew replied:
Thanks! I'll fix this up soon and re-request you when I get it done

victor-a-a replied:
Addressed in commit `678345a`

victor-a-a replied:
Makefile addressed in commit `bf4826f`

victor-a-a commented on src/wdl/Makefile:
```diff
@@ -0,0 +1,49 @@
+clean-bin:
+	-${RM} ${LIB} ${OBJS} $(SRCS:.c=.d) $(EXAMPLE_SRCS:.c=.d) $(EXAMPLE_BINS)
+
+clean: clean-lib clean-lib
```
Makefile addressed in commit `bf4826f`

Kate32815 commented on src/wdl/include/validate.h:
```diff
@@ -0,0 +1,120 @@
+ *
+ * parameters:
+ *  - ls: a list of objects
+ *  - a function pointer to a type checking function
```
This hasn't been corrected yet, should be “function pointer to a type object printing function

mewthew replied:
Addressed in commit 2df2034

Kate32815 commented on src/wdl/src/validate.c:
```diff
@@ -0,0 +1,358 @@
+// The following functions regard room type checking
+
+/* conditions_get_list()
+ * a helper function for connection_type_check that gets a list of connections
```
Should be "that gets a list of conditions associated with a connection for a room object"

mewthew replied:
Addressed in commit 2df2034

Kate32815 commented on src/wdl/include/parse.h:
```diff
@@ -0,0 +1,54 @@
+ * gets the ids of a list of room, item, or player objects
+ *
+ * parameters:
+ *  - ls: a list of room,s items, or players
```
small typo! Should be "a list of rooms, items, or players"

mewthew replied:
Addressed in commit 2df2034

thoranderskow commented on docs/wdl.md:
```diff
@@ -0,0 +1,260 @@
+
+#### A WDL file is a YAML file that has at least the three file objects: Game, Room, Item.
+
+### Refer to Action Management's file: src/action_management/include/actionmanagement.h on the actions/dev branch for a list of approved actions in a game
```
Addressed wdl.md in commit 4936e01

thoranderskow commented on docs/wdl.md:
```diff
@@ -0,0 +1,260 @@
+
+        state: `<STRING_ADJ>` which is the descriptor for the state of the item
+
+        value: `<VAL>` which is the value of the state of the item upon initializaition of the game
```
Addressed wdl.md in commit 4936e01

thoranderskow commented on docs/wdl.md:
```diff
@@ -0,0 +1,260 @@
+
+    state: `<STRING_ADJ>` which is the descriptor for the state of the item
+    
+    value: `<VAL>` which is the value of the state of the item upon initializaition of the game
```
Addressed wdl.md in commit 4936e01

thoranderskow commented on docs/wdl.md:
```diff
@@ -0,0 +1,260 @@
+
+      text_success: `<STRING>` which is a string that is displayed upon the success of an action (OPTIONAL)
+
+      text_fail: `<STRING>` which is the string that is displayed when an action is not allowed
```
Addressed wdl.md in commit 1ced07f. Upon review, it appears that you should be able to make an action always succeed if attempted, thus text_fail was made optional.

thoranderskow commented on docs/wdl.md:
```diff
@@ -0,0 +1,260 @@
+      
+      - id: "star"
+
+        state: "in inventory"
```
Discussed with Borja, this will be addressed in the next sprint.

thoranderskow replied:
Addressed wdl.md by removing state and value in commit 67db93b

Kate32815 commented on src/wdl/include/validate.h:
```diff
@@ -80,7 +80,7 @@ bool list_type_check(attr_list_t *ls, bool(*validate)(obj_t*));
  * parameters:
  *  - ls: a list of objects
- *  - a function pointer to a type checking function
+ *  - a function pointer to a type object printing function
```
should be "- a function pointer to an object printing function" (remove "type")

mewthew replied:
fixed! in commit e50e61f

lydiafilipe commented:
Issue Score: ✓

Comments:
Lots of good work here! However you are missing a closing summary and the title isn't particularly descriptive.

<!-- {"score": "check"} -->
